chain/client/src/stateless_validation/state_witness_tracker.rs

```diff
@@ -0,0 +1,212 @@
+/// the originating witness record.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+struct ChunkStateWitnessKey {
+    /// Hash of the chunk for which the state witness was generated.
+    chunk_hash: ChunkHash,
+}
+
+impl ChunkStateWitnessKey {
+    pub fn new(witness: &ChunkStateWitness) -> Self {
+        Self { chunk_hash: witness.inner.chunk_header.chunk_hash() }
+    }
+}
+
+struct ChunkStateWitnessRecord {
+    /// Size of the witness in bytes.
+    witness_size: usize,
+    /// Number of validators that the witness is sent to.
+    num_validators: usize,
+    /// Timestamp of when the chunk producer sends the state witness.
+    sent_timestamp: near_async::time::Instant,
+}
+
+/// Tracks a collection of state witnesses sent from chunk producers to validators.
+///
+/// This is currently used to calculate the round-trip time of sending the witness and
+/// getting the ack message back, where the ack message is used as a proxy for the endorsement
+/// message from validators to block producers to make an estimate of network time for sending
+/// witness and receiving the endorsement.
+pub struct ChunkStateWitnessTracker {
+    witnesses: LruCache<ChunkStateWitnessKey, ChunkStateWitnessRecord>,
+    clock: Clock,
+}
+
+impl ChunkStateWitnessTracker {
+    pub fn new(clock: Clock) -> Self {
+        Self { witnesses: LruCache::new(CHUNK_STATE_WITNESS_MAX_RECORD_COUNT), clock }
+    }
+
+    /// Adds a new witness message to track.
+    pub fn record_witness_sent(
+        &mut self,
+        witness: &ChunkStateWitness,
+        witness_size_in_bytes: usize,
+        num_validators: usize,
+    ) -> () {
+        let key = ChunkStateWitnessKey::new(witness);
+        tracing::trace!(target: "state_witness_tracker", witness_key=?key,
+            size=witness_size_in_bytes, "Recording state witness sent.");
+        self.witnesses.put(
+            key,
+            ChunkStateWitnessRecord {
+                num_validators,
+                witness_size: witness_size_in_bytes,
+                sent_timestamp: self.clock.now(),
+            },
+        );
+    }
+
+    /// Handles an ack message for the witness. Calculates the round-trip duration and
+    /// records it in the corresponding metric.
+    pub fn on_witness_ack_received(&mut self, ack: ChunkStateWitnessAck) -> () {
+        let key = ChunkStateWitnessKey { chunk_hash: ack.chunk_hash };
+        tracing::trace!(target: "state_witness_tracker", witness_key=?key,
+            "Received ack for state witness");
+        if let Some(record) = self.witnesses.get_mut(&key) {
+            debug_assert!(record.num_validators > 0);
+
+            Self::update_roundtrip_time_metric(record, &self.clock);
+
+            // Cleanup the record if we received the acks from all the validators, otherwise update
+            // the number of validators from which we are expecting an ack message.
+            match record.num_validators.checked_sub(1) {
```

Comment:
this is completely is fine but if you don't want to deal with options you can use saturating_sub instead

Reply:
much better. thanks.